ambry-api/src/test/java/com.github.ambry/account/InMemAccountServiceFactory.java

```diff
@@ -0,0 +1,198 @@
+
+  private final boolean returnOnlyUnknown;
+
+  /**
+   * Constructor. If the properties contains a field "in.mem.account.service.only.unknown" set to {@code true}, an
+   * {@link InMemAccountService} that only returns {@link Account#UNKNOWN_ACCOUNT} is returned. Otherwise a fully
+   * functional service is returned. These account services are also static (singleton) so the same instance of these
+   * services is returned no matter how many times {@link #getAccountService()} is called or different instances of
+   * {@link InMemAccountServiceFactory} are created.
+   * @param verifiableProperties The properties to get a {@link InMemAccountService} instance. Cannot be {@code null}.
+   * @param metricRegistry will be discarded
+   * @param notifier will be discarded
+   */
+  public InMemAccountServiceFactory(VerifiableProperties verifiableProperties, Object metricRegistry, Object notifier) {
+    returnOnlyUnknown = verifiableProperties.getBoolean("in.mem.account.service.only.unknown", false);
+  }
+
+  public InMemAccountServiceFactory(boolean returnOnlyUnknown) {
+    this.returnOnlyUnknown = returnOnlyUnknown;
+  }
+
+  @Override
+  public InMemAccountService getAccountService() {
+    return returnOnlyUnknown ? ONLY_UNKNOWN : ANY_ACCOUNT;
+  }
+
+  /**
+   * Implementation of {@link AccountService} for test. This implementation synchronizes on all methods.
+   */
+  public static class InMemAccountService implements AccountService {
+    private final boolean shouldReturnOnlyUnknownAccount;
+    private final Map<Short, Account> idToAccountMap = new HashMap<>();
+    private final Map<String, Account> nameToAccountMap = new HashMap<>();
+    private final Set<Consumer<Collection<Account>>> accountUpdateConsumers = new HashSet<>();
+
+    /**
+     * Constructor.
+     * @param shouldReturnOnlyUnknownAccount {@code true} if always returns {@link Account#UNKNOWN_ACCOUNT} when queried
+     *                                                     by account name; {@code false} to do the actual query.
+     */
+    private InMemAccountService(boolean shouldReturnOnlyUnknownAccount) {
+      this.shouldReturnOnlyUnknownAccount = shouldReturnOnlyUnknownAccount;
+    }
+
+    @Override
+    public synchronized Account getAccountById(short accountId) {
+      return shouldReturnOnlyUnknownAccount ? Account.UNKNOWN_ACCOUNT : idToAccountMap.get(accountId);
+    }
+
+    @Override
+    public synchronized Account getAccountByName(String accountName) {
+      return shouldReturnOnlyUnknownAccount ? Account.UNKNOWN_ACCOUNT : nameToAccountMap.get(accountName);
+    }
+
+    @Override
+    public synchronized boolean updateAccounts(Collection<Account> accounts) {
+      for (Account account : accounts) {
+        idToAccountMap.put(account.getId(), account);
+        nameToAccountMap.put(account.getName(), account);
+      }
+      for (Consumer<Collection<Account>> accountUpdateConsumer : accountUpdateConsumers) {
```

Comment:
small ask: could you add a `shouldNotifyConsumers` flag/setter? This would help with some of my test cases. Basically, an option to disable notifying accountUpdateConsumers on updates when set to false.

Reply:
Done